Blender: Fix Deadline Frames per task

simonebarbieri opened the pull request:
## Changelog Description
Fixed a problem with Frames per task setting not being applied when publishing a render.

## Additional info
This PR also adds the attributes in the publishing screen. ~~However, this required to change the family of the render instance, so it is necessary to recreate any existing render instance.~~ Fixed
This PR adds the Delay option as well.

## Testing notes:
1. Recreate any existing render instance, or create one in a scene without any.
2. Set in the publishing screen any number in the "Frame per task" option.
3. Verify in Deadline that the option has been applied.

BigRoy commented:
Not sure how urgent this one is but I have some notes on it - but not behind a PC right now. Give me a few hours to comment. :)

BigRoy commented on openpype/hosts/blender/api/render_lib.py:
```diff
@@ -59,7 +59,7 @@ def get_render_product(output_path, name, aov_sep):
         ext (str): The image format to render.
     """
-    filepath = os.path.join(output_path, name)
+    filepath = output_path / name.relative_to(name.anchor)
```
`name.relative_to(name.anchor)` 

This could easily introduce massive paths, right? :) 

The anchor being e.g. `c:/` and the path being `c:/project/asset/work/lighting/render/render.0001.png` basically appending `project/asset/work/lighting/render/render.0001.png` to the output dir?

Or why are we doing it like this?

I imagine we want to make this relative maybe to the render root directory?

simonebarbieri replied:
This actually solves a bug that I missed when testing the PR to use the folder paths instead of the asset name 😅 The problem is that the new asset starts with `/`, so when using `os.path.join` it takes it as an anchor and breaks the path, that's why it is necessary! The path is the same as it was before (which is basically a subdir of the workfile) :)

BigRoy replied:
Might the intention be clearer if we do `name.lstrip("/")`

BigRoy commented on openpype/hosts/blender/api/render_lib.py:
```diff
@@ -232,11 +232,11 @@ def prepare_rendering(asset_group):
     aov_list, custom_passes = set_render_passes(settings)
 
-    output_path = os.path.join(file_path, render_folder, file_name)
+    output_path = Path.joinpath(file_path, render_folder, file_name)
```
Likely personal preference but this should also work, right?
```suggestion
    output_path = file_path / render_folder / file_name
```

simonebarbieri replied:
Yes it's the same, I like `joinpath` even if longer because it's more clear :) but I can be convinced to change it :P

LiborBatek reviewed (changes requested):
First issue being the ```render``` instance name being wrong set to ```farm``` and family to ```render.farm```
<img width="480" alt="image" src="https://github.com/ynput/OpenPype/assets/112623825/d1414454-b533-4439-b99c-8be816c64f98">

~~Also this being issue?~~ I have set my DL incorrectly now this not being issue and renders ok!

<img width="777" alt="image" src="https://github.com/ynput/OpenPype/assets/112623825/bc72afae-07cf-4464-81cb-9fd24f461fd2">

<img width="466" alt="image" src="https://github.com/ynput/OpenPype/assets/112623825/181f3698-0b24-46c9-8e85-88fc6a46d9ed">

On DL it correctly uses ```frame per task``` using my value ```10```

<img width="726" alt="image" src="https://github.com/ynput/OpenPype/assets/112623825/5d9f8e92-9cf5-48ad-870a-507d65206e83">

BigRoy commented on openpype/modules/deadline/plugins/publish/submit_blender_deadline.py:
```diff
@@ -83,8 +88,9 @@ def get_job_info(self):
 
-        job_info.Priority = attr_values.get("priority")
-        job_info.ChunkSize = attr_values.get("chunkSize")
+        job_info.Comment = context.data.get("comment")
```
Should the comment come from the instance?

There's this global Instance Per Comment plug-in that you can enable to allow custom comments per instance: https://github.com/pypeclub/OpenPype/blob/871bf194da2e7743363697032a984c41a2e54fc4/openpype/plugins/publish/collect_comment.py#L35

If not used, [this](https://github.com/ynput/OpenPype/blob/871bf194da2e7743363697032a984c41a2e54fc4/openpype/plugins/publish/collect_comment.py#L83-L104) should be transferring the `context.data["comment"]` to the instance already.

As such, by using instance here you basically make sure you also support the (optional per family) instance per comment plug-in.

Would that work?

simonebarbieri replied:
I don't know enough about this and I ~~blatantly copied~~ took inspiration from the Max submitter to implement the Blender submitter 😅
I'll try it out :)

BigRoy replied:
Might want to then push a similar fix to max if this works here

simonebarbieri replied:
It works :)

@moonyuet we should consider doing it for Max as well :)

BigRoy commented on openpype/hosts/blender/api/render_lib.py:
```diff
@@ -212,14 +213,13 @@ def imprint_render_settings(node, data):
 
 def prepare_rendering(asset_group):
-    name = asset_group.name
+    name = Path(asset_group.name)
```
Is the asset group's name a Path? Is that correct?


Also you're turning it into `Path()` here just to do str(name)` later; is it a redundant conversion?

simonebarbieri replied:
You're right 😅 there is `get_render_product` that uses it as `Path`, and `set_node_tree` that uses it as `str`! I guess it's better to just initialize it as `Path` when passing it to `get_render_product`! I'll change it!

BigRoy commented on openpype/hosts/blender/api/render_lib.py:
```diff
@@ -212,14 +213,13 @@ def imprint_render_settings(node, data):
+    filepath = Path(bpy.data.filepath)
     assert filepath, "Workfile not saved. Please save the file first."
 
-    file_path = os.path.dirname(filepath)
-    file_name = os.path.basename(filepath)
-    file_name, _ = os.path.splitext(file_name)
+    file_path = filepath.parent
```
Having both `filepath` and file_path` variable in this scope feels a bit confusing. Should this be `filepath = filepath.parent` or `directory = filepath.parent`?

The parent isn't technically a filepath, it's the path to a directory/folder instead.

simonebarbieri replied:
Yep you're right! I'll change it :)

BigRoy commented:
> However, this required to change the family of the render instance, so it is necessary to recreate any existing render instance.

Just a side note: This actually shows a different bug - the `family` should be defined by the creator, not by a `family` attribute on the instance in the scene. The instance in the scene knows which creator it belongs to; so next time it runs it should be able to identify using the right family. (The subset familly might then technically be wrong; but that's ok to persist for backwards compatibilities).

LiborBatek commented:
The final render got published correctly and without any issues...

<img width="870" alt="image" src="https://github.com/ynput/OpenPype/assets/112623825/2fe6728f-b7b9-4c23-93af-f74a7a0ce9b1">

LiborBatek commented:
however when trying to recreate ```Render``` instance Im getting this error when trying in Publisher:

```Creator: Render (io.openpype.creators.blender.render)

Error: 'str' object has no attribute 'relative_to'

Traceback (most recent call last):
  File "C:\Work\REPO\OpenPype\openpype\pipeline\create\context.py", line 2059, in _create_with_unified_error
    result = creator.create(*args, **kwargs)
  File "C:\Work\REPO\OpenPype\openpype\hosts\blender\plugins\create\create_render.py", line 25, in create
    prepare_rendering(collection)
  File "C:\Work\REPO\OpenPype\openpype\hosts\blender\api\render_lib.py", line 238, in prepare_rendering
    aov_file_product = set_node_tree(
  File "C:\Work\REPO\OpenPype\openpype\hosts\blender\api\render_lib.py", line 194, in set_node_tree
    filename = output_path / filepath.relative_to(filepath.anchor)
AttributeError: 'str' object has no attribute 'relative_to'
```

LiborBatek commented:
> however when trying to recreate `Render` instance Im getting this error when trying in Publisher:
> 
> ```
> 
> Error: 'str' object has no attribute 'relative_to'
> 
> Traceback (most recent call last):
>   File "C:\Work\REPO\OpenPype\openpype\pipeline\create\context.py", line 2059, in _create_with_unified_error
>     result = creator.create(*args, **kwargs)
>   File "C:\Work\REPO\OpenPype\openpype\hosts\blender\plugins\create\create_render.py", line 25, in create
>     prepare_rendering(collection)
>   File "C:\Work\REPO\OpenPype\openpype\hosts\blender\api\render_lib.py", line 238, in prepare_rendering
>     aov_file_product = set_node_tree(
>   File "C:\Work\REPO\OpenPype\openpype\hosts\blender\api\render_lib.py", line 194, in set_node_tree
>     filename = output_path / filepath.relative_to(filepath.anchor)
> AttributeError: 'str' object has no attribute 'relative_to'
> ```

this seems to happen when blender missing some ```addon```?? when fixed and placed the needed addon into ```scripts/addons``` it does not show up again!

LiborBatek commented:
we have another issue that when ```Publish``` of the ```Render``` happens it wont upversion the workfile so the DL jobs end up in the same location/job entry...

<img width="958" alt="image" src="https://github.com/ynput/OpenPype/assets/112623825/e1e1010f-e595-4b33-8f86-d5cfccc5df19">

simonebarbieri commented:
> Just a side note: This actually shows a different bug - the family should be defined by the creator, not by a family attribute on the instance in the scene. The instance in the scene knows which creator it belongs to; so next time it runs it should be able to identify using the right family. (The subset familly might then technically be wrong; but that's ok to persist for backwards compatibilities).

@BigRoy yeah the problem was that the instance was empty because the collector I was using for `render` it now works for `render.farm`. But I didn't think that I can actually leave the `render` family in the collector together with `render.farm` and it works 😅 it just misses the options in the publisher, but at least it works with the options from the settings :) I'll commit the fix in a second

simonebarbieri commented:
> First issue being the render instance name being wrong set to farm and family to render.farm

@LiborBatek That's was only way (that I found) to add the options in the publisher :/ the problem is that `BlenderSubmitDeadline` and `ProcessSubmittedJobOnFarm` work on `render.farm` families, and I cannot add `render` to `ProcessSubmittedJobOnFarm` because it might pick up wrong instances in other clients. The only solution is to use `render.farm` (which isn't wrong, in my opinion, as this render family is very specific for submitting to Deadline) in Blender, or add a new family (like Max does with `maxrender`) that I can add to `ProcessSubmittedJobOnFarm` without messing up other DCCs 😅

simonebarbieri commented:
> however when trying to recreate Render instance Im getting this error when trying in Publisher:
> Updated info: this issue being connected to when create AOV set via OP Settings

@LiborBatek I should have fixed it, could you try it again? :)

antirotor commented:
I don't know. We are touching the limitation of the current Publisher that it can't resolve final order of plugins that will run until the collection phase so we can't show attributes of plugins that will run only after some collector will add their family.

In nuke, deadline submitter plugin runs on `render` family and only in `process()` method at the beginning it checks if there is `farm` or not. Disadvantage of it is that this plugin is always executed no matter if it render should run on farm or not. Similar approach is done in Maya except there is no local rendering supported (yet) - it is run for `renderlayer` family.

In Houdini, it is using again the family set by creator.

In Fusion, `render` is the primary family and `render.local` is added for those you'd like to render locally - extractor is ignoring any `render` instances that miss `render.local`.

I suggest we do something similar here, `farm` isn't good primary family for renders.

BigRoy commented:
> I don't know. We are touching the limitation of the current Publisher that it can't resolve final order of plugins that will run until the collection phase so we can't show attributes of plugins that will run only after some collector will add their family.

I this is the only thing that you need solved then have a look at my PRs that implement `get_publish_families` on the Creators. That way your Creator can just specify:

```python
class CreateRender:
    def get_publish_families(self):
        return ["render.farm", "render"]
```

However, it'll always have both those families for that creator - it's not depending on a specific attribute being set or not.

But I suppose since `farm` is a toggle on the instance itself that might still require what is being done in Nuke/Maya as well where the extractors ignore specific instance data, etc.

simonebarbieri commented:
> In Fusion, `render` is the primary family and `render.local` is added for those you'd like to render locally - extractor is ignoring any `render` instances that miss `render.local`.

@antirotor I am checking Fusion, and I can see that it is using `render` as family. I've tried this with Blender, but because `ProcessSubmittedJobOnFarm` is missing `render` family (and I am afraid to add it because I think it would break DCCs that use `render` for local rendering), it misses the Ayon task in the Deadline job. How is this solved in Fusion?

simonebarbieri commented:
I've fixed the family to be `render` again, and fixed the problem with `ProcessSubmittedJobOnFarm` missing the ayon/openpype task in deadline!

This time, the instances created as `farm` while testing this PR won't work, so you need to recreate them again or use the previous `render` instances you had 😅

LiborBatek commented:
Now rendering works in OP legacy and rendered sequences got correctly produced but I cant see any trace of my ```delay time``` attribute in action.

It just starts immedeately rendering and not waiting for ```delay``` value set in OP Settings. Also checked output log and ```Job Properties``` in DL but no any trace of use of it.

Will inspect it also in Ayon mode in while...

LiborBatek reviewed (changes requested):
Neither in OP legacy nor Ayon mode, the ```delay time``` attrib is not taken into account when rendering on DL.

Need to be addressed first. Regarding the actual rendering of the sequences all seems normal and working predictably.

BigRoy commented on openpype/modules/deadline/plugins/publish/submit_blender_deadline.py:
```diff
@@ -180,3 +188,36 @@ def from_published_scene(self):
+                    default=cls.group,
+                    label="Group Name"),
+
+            TextDef("job_delay",
+                    default=cls.job_delay,
+                    label="Job Delay"),
```
This could benefit from placeholder text: `dd:hh:mm:ss` just so it's clear what the user _should_ enter there if they were to ever clear it. I'd also say that adding that As a tooltip would be helpful as well.

The tooltip then also explain that the delay is added to the submission time.

Basically the tooltip could match the label in settings: `"Delay job (timecode dd:hh:mm:ss)"`

---

Out of curiosity, from where comes the request to "delay" the render?

LiborBatek replied:
> 
> Out of curiosity, from where comes the request to "delay" the render?

I guess thats because of case when syncing data from cloud storage like Google / AWS taking some time to sync so u need some time for that in advance delaying the render job.

BigRoy replied:
But a short delay wouldn't be safe then though since you wouldn't _really_ know when the sync might be done? 

To wait for dependencies you should set up your job with `assetDependencies` so that Deadline only starts once those dependencies are found by deadline? Or? I believe that's what it was designed for at least.

LiborBatek replied:
> But a short delay wouldn't be safe then though since you wouldn't _really_ know when the sync might be done?
> 
> To wait for dependencies you should set up your job with `assetDependencies` so that Deadline only starts once those dependencies are found by deadline? Or? I believe that's what it was designed for at least.

I do agree, but this being sort of simple workaround/way to do it too... similar being already done when speaking of DL integration for Celaction host.  This also comes from users needs aka they had been already using it before... and now not being possible/ option.

BigRoy replied:
Thanks - was just wondering about the request. Didn't know about the feature - thanks for the contribution @simonebarbieri ❤️

m-u-r-p-h-y replied:
Some DCCs do not provide the complete list of dependent files (even Maya with Yeti for example), so it is not reliable to use just this feature. In this specific case the client uses Google Drive and using the delay for job start is just bullet proof for them to allow the render scene (and other files) to sync correctly . . .

LiborBatek reviewed (approved):
Did final round of tests on AYON and all working well as in legacy ver.

Approved.
